grails-app/controllers/au/org/ala/specieslist/SpeciesListController.groovy

```diff
@@ -340,32 +363,77 @@ class SpeciesListController {
     /**
      * Performs an initial parse of the species list to provide feedback on values. Allowing
      * users to supply vocabs etc.
+     *
+     * Data can be submitted either via a file upload or as copy and paste text
      */
-    def parseData(){
+    def parseData() {
         log.debug("Parsing for header")
-        def rawData = request.getReader().readLines().join("\n").trim()
-        String separator = helperService.getSeparator(rawData)
-        CSVReader csvReader = helperService.getCSVReaderForText(rawData, separator)
-        def rawHeader =  csvReader.readNext()
+
+        String separator
+        CSVReader csvReader
+        CommonsMultipartFile file = isMultipartRequest() ? request.getFile(CSV_UPLOAD_FILE_NAME) : null
+
+        if (file) {
+            if (ACCEPTED_CONTENT_TYPES.contains(file.getContentType())) {
+                separator = detectSeparator(file);
+                csvReader = helperService.getCSVReaderForCSVFileUpload(file, separator as char)
+            } else {
+                render(view: 'parsedData', model: [error: INVALID_FILE_TYPE_MESSAGE])
+                return
+            }
+        } else {
+            def rawData = request.getReader().readLines().join("\n").trim()
+            separator = helperService.getSeparator(rawData)
+            csvReader = helperService.getCSVReaderForText(rawData, separator)
+        }
+
+        try {
+            parseDataFromCSV(csvReader, separator)
+        }
+        finally {
+            csvReader?.close()
+        }
+    }
+
+    private String detectSeparator(CommonsMultipartFile file) {
+        String separator
+        def reader
+        try {
+            reader = new BufferedReader(new InputStreamReader(file.getInputStream()))
+            separator = helperService.getSeparator(reader.readLine())
+        }
+        finally {
+            reader?.close()
+            // reset the uploaded file's input stream so that it can be read from again (by the CSV reader)
+            file.getInputStream().reset()
+        }
+
+        separator
+    }
+
+    private parseDataFromCSV(CSVReader csvReader, String separator) {
+        def rawHeader = csvReader.readNext()
         log.debug(rawHeader.toList())
-        def parsedHeader = helperService.parseHeader(rawHeader)?:helperService.parseData(rawHeader)
+        def parsedHeader = helperService.parseHeader(rawHeader) ?: helperService.parseData(rawHeader)
         def processedHeader = parsedHeader.header
         log.debug(processedHeader)
         def dataRows = new ArrayList<String[]>()
         def currentLine = csvReader.readNext()
-        for(int i = 0; i < noOfRowsToDisplay && currentLine != null; i++){
+        for (int i = 0; i < noOfRowsToDisplay && currentLine != null; i++) {
             dataRows.add(currentLine)
```

Comment:
CSVReader is used all through the application - this was existing code. I agree that it could be refactored, but that is beyond the scope of the current task. I think we should leave this for now.